src/common/src/util/sync.rs

```diff
@@ -0,0 +1,160 @@
+        }
+    }
+
+    async fn wait_for_signal(&self, wait_for_signal: WaitForSignal) -> Result<(), Error> {
+        let entry = self
+            .signals
+            .lock()
+            .entry(wait_for_signal.signal.to_owned())
+            .or_insert_with(|| Arc::new(tokio::sync::Notify::new()))
+            .clone();
+        match tokio::time::timeout(wait_for_signal.timeout, entry.notified()).await {
+            Ok(_) => {
+                if !wait_for_signal.clear_signal {
+                    entry.notify_one();
+                }
+            }
+            Err(_) => {
+                return Err(Error::WaitForSignalTimeout(wait_for_signal.signal));
+            }
+        }
+        Ok(())
+    }
+
+    fn emit_signal(&self, signal: Signal) {
+        let entry = self
+            .signals
+            .lock()
+            .entry(signal)
+            .or_insert_with(|| Arc::new(tokio::sync::Notify::new()))
+            .clone();
+        entry.notify_one();
+    }
+
+    fn set_actions(&self, sync_point: &str, actions: Vec<Action>, execute_times: u64) {
+        let mut guard = self.sync_points.lock();
+        let sync_points = guard.deref_mut();
+        sync_points.insert(
+            sync_point.to_owned(),
+            SyncPointInfo {
+                actions,
+                execute_times,
+            },
+        );
+    }
+
+    fn reset_actions(&self, sync_point: &str) {
+        self.sync_points.lock().remove(sync_point);
+    }
+
+    async fn on_sync_point(&self, sync_point: &str) -> Result<(), Error> {
+        let actions = {
+            let mut guard = self.sync_points.lock();
+            match guard.entry(sync_point.to_owned()) {
+                Entry::Occupied(mut o) => {
+                    if o.get().execute_times <= 1 {
+                        // Deactivate the sync point
+                        guard.remove(sync_point).unwrap().actions
+                    } else {
+                        o.get_mut().execute_times -= 1;
+                        o.get().actions.clone()
+                    }
+                }
+                Entry::Vacant(_) => {
+                    return Ok(());
+                }
+            }
+        };
+        for action in actions {
+            match action {
+                Action::WaitForSignal(w) => {
+                    self.wait_for_signal(w.to_owned()).await?;
+                }
+                Action::EmitSignal(s) => {
+                    self.emit_signal(s.to_owned());
+                }
+            }
+        }
+        Ok(())
+    }
+}
+
+/// The activation is reset after executed `execute_times`.
+pub fn activate_sync_point(sync_point: &str, actions: Vec<Action>, execute_times: u64) {
+    SYNC_FACILITY.set_actions(sync_point, actions, execute_times);
+}
+
+pub fn deactivate_sync_point(sync_point: &str) {
+    SYNC_FACILITY.reset_actions(sync_point);
+}
+
+pub async fn on_sync_point(sync_point: &str) -> Result<(), Error> {
```

Comment:
maybe to make this zero-cost for non-test, we can either use #cfg to conditionally compile, or inline this and hope the compiler removes the branch (I prefer the former)

Reply:
Fixed.